File: 2023/day03/c/solution.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <stdbool.h>

#define MAX_ROWS 200
#define MAX_COLS 200

char grid[MAX_ROWS][MAX_COLS];
int rows = 0;
int cols = 0;

// Structure to track a number and its position
typedef struct {
    int value;
    int row;
    int start_col;
    int end_col;
} Number;

Number numbers[5000];
int num_count = 0;

// Check if a character is a symbol (not digit, not period)
bool is_symbol(char c) {
    return c != '.' && !isdigit(c) && c != '\0' && c != '\n';
}

// Check if a number at given position is adjacent to any symbol
bool is_adjacent_to_symbol(int row, int start_col, int end_col) {
    // Check all 8 directions around the number
    for (int r = row - 1; r <= row + 1; r++) {
        for (int c = start_col - 1; c <= end_col + 1; c++) {
            // Skip out of bounds
            if (r < 0 || r >= rows || c < 0 || c >= cols) continue;
            // Skip the number itself
            if (r == row && c >= start_col && c <= end_col) continue;

            if (is_symbol(grid[r][c])) {
                return true;
            }
        }
    }
    return false;
}

// Get all numbers adjacent to a position
void get_adjacent_numbers(int row, int col, int *adj_nums, int *count) {
    *count = 0;

    // Check all 8 directions
    for (int r = row - 1; r <= row + 1; r++) {
        for (int c = col - 1; c <= col + 1; c++) {
            // Skip out of bounds
            if (r < 0 || r >= rows || c < 0 || c >= cols) continue;
            // Skip center
            if (r == row && c == col) continue;

            // If this is a digit, find the complete number
            if (isdigit(grid[r][c])) {
                // Find which number this belongs to
                for (int i = 0; i < num_count; i++) {
                    if (numbers[i].row == r &&
                        c >= numbers[i].start_col &&
                        c <= numbers[i].end_col) {
                        // Check if we already have this number index
                        bool found = false;
                        for (int j = 0; j < *count; j++) {
                            if (adj_nums[j] == i) {
                                found = true;
                                break;
                            }
                        }

                        if (!found) {
                            adj_nums[*count] = i;  // Store index
                            (*count)++;
                        }
                        break;
                    }
                }
            }
        }
    }
}

int part1() {
    int sum = 0;

    for (int i = 0; i < num_count; i++) {
        if (is_adjacent_to_symbol(numbers[i].row, numbers[i].start_col, numbers[i].end_col)) {
            sum += numbers[i].value;
        }
    }

    return sum;
}

int part2() {
    int sum = 0;

    // Find all '*' symbols
    for (int r = 0; r < rows; r++) {
        for (int c = 0; c < cols; c++) {
            if (grid[r][c] == '*') {
                int adj_nums[10];
                int count;
                get_adjacent_numbers(r, c, adj_nums, &count);

                // If exactly 2 adjacent numbers, it's a gear
                if (count == 2) {
                    int gear_ratio = numbers[adj_nums[0]].value * numbers[adj_nums[1]].value;
                    sum += gear_ratio;
                }
            }
        }
    }

    return sum;
}

int main(int argc, char *argv[]) {
    const char *filename = (argc > 1) ? argv[1] : "../input.txt";
    FILE *f = fopen(filename, "r");
    if (!f) {
        perror("Error opening file");
        return 1;
    }

    // Read the grid
    char line[MAX_COLS];
    while (fgets(line, sizeof(line), f) && rows < MAX_ROWS) {
        // Remove newline
        line[strcspn(line, "\n")] = 0;
        strcpy(grid[rows], line);
        if (cols == 0) {
            cols = strlen(line);
        }
        rows++;
    }
    fclose(f);

    // Parse all numbers and their positions
    for (int r = 0; r < rows; r++) {
        for (int c = 0; c < cols; c++) {
            if (isdigit(grid[r][c])) {
                int start = c;
                int value = 0;

                // Parse the complete number
                while (c < cols && isdigit(grid[r][c])) {
                    value = value * 10 + (grid[r][c] - '0');
                    c++;
                }

                // Store the number
                numbers[num_count].value = value;
                numbers[num_count].row = r;
                numbers[num_count].start_col = start;
                numbers[num_count].end_col = c - 1;
                num_count++;

                c--; // Adjust because loop will increment
            }
        }
    }

    printf("Part 1: %d\n", part1());
    printf("Part 2: %d\n", part2());

    return 0;
}

File: 2023/day03/clojure/solution.clj
(ns solution
  (:require [clojure.string :as str]))

(def input
  (-> (slurp "../input.txt")
      str/trim
      (str/split #"\n")))

(defn char-at [grid row col]
  "Get character at position, returns nil if out of bounds"
  (when (and (>= row 0) (< row (count grid))
             (>= col 0) (< col (count (nth grid row))))
    (get-in grid [row col])))

(defn is-symbol? [ch]
  "Check if character is a symbol (not digit, not period, not nil)"
  (and ch
       (not= ch \.)
       (not (Character/isDigit ch))))

(defn find-numbers [grid]
  "Find all numbers in the grid with their positions.
   Returns a sequence of maps with :value, :row, :start-col, :end-col"
  (for [row (range (count grid))
        :let [line (nth grid row)
              matches (re-seq #"\d+" line)]
        :when (seq matches)]
    (loop [col 0
           remaining matches
           results []]
      (if (empty? remaining)
        results
        (let [num-str (first remaining)
              num-val (parse-long num-str)
              idx (str/index-of line num-str col)]
          (recur (+ idx (count num-str))
                 (rest remaining)
                 (conj results {:value num-val
                               :row row
                               :start-col idx
                               :end-col (+ idx (count num-str) -1)})))))))

(defn get-adjacent-positions [num-info]
  "Get all adjacent positions (including diagonals) for a number"
  (let [{:keys [row start-col end-col]} num-info]
    (for [r (range (dec row) (+ row 2))
          c (range (dec start-col) (+ end-col 2))
          :when (not (and (= r row)
                         (>= c start-col)
                         (<= c end-col)))]
      [r c])))

(defn adjacent-to-symbol? [grid num-info]
  "Check if a number is adjacent to any symbol"
  (some (fn [[r c]]
          (is-symbol? (char-at grid r c)))
        (get-adjacent-positions num-info)))

(defn part1 []
  (let [numbers (flatten (find-numbers input))
        part-numbers (filter #(adjacent-to-symbol? input %) numbers)]
    (reduce + 0 (map :value part-numbers))))

(defn find-gears [grid]
  "Find all * symbols and their adjacent numbers.
   Returns gear ratios for gears with exactly 2 adjacent numbers."
  (let [numbers (flatten (find-numbers grid))
        gear-positions (for [row (range (count grid))
                            col (range (count (nth grid row)))
                            :when (= (char-at grid row col) \*)]
                        [row col])]
    (for [[gr gc] gear-positions
          :let [adjacent-numbers
                (filter (fn [num-info]
                         (some (fn [[r c]] (and (= r gr) (= c gc)))
                               (get-adjacent-positions num-info)))
                       numbers)]
          :when (= (count adjacent-numbers) 2)]
      (apply * (map :value adjacent-numbers)))))

(defn part2 []
  (reduce + 0 (find-gears input)))

;; Main execution
(println "Part 1:" (part1))
(println "Part 2:" (part2))
